### PR TITLE
supervisord does not expand shell variables, we need to call out to a shell

### PR DESCRIPTION
**Pull request by Benjamin-Dobell**
I'm not sure why, but:

```
-command=/exec sh -c '${COMMAND}'
+command=/exec ${COMMAND}
```

Was included as part of https://github.com/sehrope/dokku-logging-supervisord/commit/70f5ea273ec6dc5d2f809315fed6985779ecd913. This is an incorrect change as supervisord does not perform shell expansion of variables included in COMMAND.

So something simple like:

```
bundle exec rackup -p $PORT
```

fails with the current behaviour.

### CODE REVIEW

**Comment by michaelshobbs** [comment]
This doesn't seem to work with the latest buildstep release...
Example with this repo: https://github.com/heroku/node-js-sample

Procfile (built by nodejs buildpack):
```
web: npm start
```

Logs:
```
sh: 0: -c requires an argument
```

supervisord.conf
```
Generated supervisord.conf:
            1	[supervisord]
            2	logfile=/var/log/app/supervisor.log
            3	loglevel=info
            4	nodaemon=true
            5
            6	[program:web]
            7	command=/exec sh -c "${COMMAND}"
            8	process_name:%(program_name)s
            9	numprocs=1
           10	autostart=true
           11	autorestart=true
           12	stopsignal=TERM
           13	redirect_stderr=true
           14	stdout_logfile=/var/log/app/%(program_name)s.%(process_num)02d.log
```

**Comment by Benjamin-Dobell** [comment]
Sorry, I did a sneaky force push in there - was missing backslash. Could you please double check you grabbed the latest commit?

**Comment by michaelshobbs** [comment]
That was with the latest:

```
function gen_supervisor_proc_conf () {
NAME="$1"
COMMAND="$2"
NUM_PROCS="$3"
PROCESS_NAME='%(program_name)s'
if [ "$NUM_PROCS" -gt 1 ]; then
  PROCESS_NAME='%(program_name)s.%(process_num)02d'
fi

cat << EOF
[program:${NAME}]
command=/exec sh -c "\${COMMAND}"
process_name:${PROCESS_NAME}
numprocs=${NUM_PROCS}
autostart=true
autorestart=true
stopsignal=TERM
redirect_stderr=true
stdout_logfile=/var/log/app/%(program_name)s.%(process_num)02d.log
EOF
}
```

**Comment by Benjamin-Dobell** [comment]
Okay, thank you for testing.

Looks like I must have been a bit hasty trying to submit this before going to bed. I'll do some testing and submit an updated patch.

**Comment by Benjamin-Dobell** [comment]
TL;DR version - My original commit (without the backslash) was/is correct.

Well, that's what I get for looking at other people's source code and letting it confuse me when I'm tired:
https://github.com/statianzo/dokku-supervisord/blob/master/post-build-buildstep#L46

In the above case they're writing a shell script, which is then later executed, hence escaping the ```${command}``` whilst the shell script is being generated (because ```command``` is not yet defined).

In our case this is the "pre-generated" script in the ```lib/``` dir.

**Comment by Benjamin-Dobell** [comment]
Sorry, to clarify - it's only correct in-so-far as the command at least executes. I've just verified that:

```dokku redeploy <app>```

doesn't work - I believe this is because we're sending the wrong signal.

```
ff738b09b6c3        dokku/genghis:latest         "/start web"        38 seconds ago      Up 37 seconds                           high_hopper            
a9db2e2bd771        dokku/genghis:latest         "/start web"        9 minutes ago       Up 9 minutes                            furious_jones
```

**Comment by Benjamin-Dobell** [comment]
Doh, I suck at this!

I patch, does work, and does shutdown the old container just fine - I forgot Dokku keeps the old container running for 60 seconds (as you can see above I'd only waited 38 seconds).

**Comment by michaelshobbs** [comment]
Are you certain you are testing this against either the latest release of buildstep and/or built from master?

This still fails on the simple nodejs app
```
Usage: npm <command>

where <command> is one of:
    access, add-user, adduser, apihelp, author, bin, bugs, c,
    cache, completion, config, ddp, dedupe, deprecate, dist-tag,
    dist-tags, docs, edit, explore, faq, find, find-dupes, get,
    help, help-search, home, i, info, init, install, issues, la,
    link, list, ll, ln, login, logout, ls, outdated, owner,
    pack, prefix, prune, publish, r, rb, rebuild, remove, repo,
    restart, rm, root, run-script, s, se, search, set, show,
    shrinkwrap, star, stars, start, stop, t, tag, test, tst, un,
    uninstall, unlink, unpublish, unstar, up, update, upgrade,
    v, verison, version, view, whoami

npm <cmd> -h     quick help on <cmd>
npm -l           display full usage info
npm faq          commonly asked questions
npm help <term>  search for help on <term>
npm help npm     involved overview

Specify configs in the ini-formatted file:
    /app/.npmrc
or on the command line via: npm <command> --key value
Config info can be viewed via: npm help config

npm@2.7.4 /app/vendor/node/lib/node_modules/npm
```

supervisord.conf
```
Generated supervisord.conf:
     1	[supervisord]
     2	logfile=/var/log/app/supervisor.log
     3	loglevel=info
     4	nodaemon=true
     5
     6	[program:web]
     7	command=/exec sh -c "npm start"
     8	process_name:%(program_name)s
     9	numprocs=1
    10	autostart=true
    11	autorestart=true
    12	stopsignal=TERM
    13	redirect_stderr=true
    14	stdout_logfile=/var/log/app/%(program_name)s.%(process_num)02d.log
```

**Comment by Benjamin-Dobell** [comment]
Sorry, I meant to provide an update.

> Are you certain you are testing this against either the latest release of buildstep and/or built from master?

No, I wasn't. I thought I was due to https://github.com/progrium/dokku/issues/1165

I will do some testing and report back. However, since updating to the latest buildstep (either the deb package, or bleeding edge herokuish implementation) I can't get my apps to run at all - with or without this change.

**Comment by Benjamin-Dobell** [comment]
Okay, this is a bug in Herokuish (and a few versions of buildstep - I haven't checked).

https://github.com/gliderlabs/herokuish/issues/30

In my case (for the ```PORT```) variable, the issue could be worked around by using:

```
sh -c "exec ${COMMAND}"
```

instead of:

```
exec sh -c "${COMMAND}"
```

The ```PORT``` variable expands fine this way and we *partially* avoid the Herokuish bug. However, it won't work for ENV variables that are loaded into the environment by the ```/exec``` command - so it's not a good long-term solution.

Best to let the Herokuish devs fix the issue and update buildstep & dokku asap.

**Comment by Benjamin-Dobell** [comment]
Now that I'm definitely using the latest buildstep I can confirm his pull request is *not* necessary using the latest buildstep - it's only "necessary" if you're using an old buildstep with the latest dokku-logging-supervisord.

Technically this change should have no real impact on execution, but the Herokuish bug means that this patch does not work with the current buildstep, so I'm closing this issue.
